### PR TITLE
fix(javascript/typescript): lambda with parens in parameters fails

### LINKED ISSUE #2189: (typescript) snippet containing a lambda which parameters are enclosed in parens fails
**Editor:** Adding a list of samples we should add tests for:

```js
const array = [1, 2, 3].reduce<number[]>((acc, next) => [...acc, next], []);
return this.sides.every((length,width=(3+2+(4/5))) => length > 0 );
const bad = ((a, b) => [...a, b]);
const bad = (a => [...a, b]);
const bad = (_ => doSomething());
const good = () => 0;
const bad = (() => 0);
```

- [x] Nested parens in expressions
- [x] () surrounded arguments
- [x] flag argument with no ()
- [x] `_` argument

Possibly relevant:
https://stackoverflow.com/questions/18703187/count-parentheses-with-regular-expression

Since to do this I think we really need to count parens in the regex that's matching => functions.

---

This is an issue I noticed downstream in VS Code itself and in the output of Asciidoctor which uses highlightjs. The snippet causing issues is this one:

````typescript
const array = [1, 2, 3].reduce<number[]>((acc, next) => [...acc, next], []);
//                             ~~~~~~~~
// Type of the initial value, in our case making it an array of numbers.
````

And gets rendered in VS Code and Asciidoctor generated HTML as follows:

image

image

When removing the arrowhead of the lambda (even though syntactically incorrect) the formatting is correct once again (since the comment at the ends gets picked up as a comment):

image

image

### PR DESCRIPTION
**Pull request by yyyc514**
- Fixes both JavaScript and TypeScript grammars

Fixes samples like:

    const bad = ((a, b) => [...a, b]);
    sides.every((length,width=(3+2+(4/5))) => length > 0 );

This is done by counting parens in the regex that finds arrows
functions. Currently we can only handle 2 levels of nesting as
shown in the second example above.

Resolves #2189.

### CODE REVIEW

**Comment by yyyc514** [comment]
Note this isn't a full parser.... this will also fall apart if say you had parenthesis inside strings inside the params... since the regex can't tell a paren inside a string from a regular one.

Always open to better suggestions.

File: src/languages/javascript.js [hunk 1]
```diff
@@ -90,6 +90,10 @@ export default function(hljs) {
     hljs.REGEXP_MODE
   ];
   var PARAMS_CONTAINS = SUBST.contains.concat([
+    // eat recursive parens in sub expressions
+    { begin: /\(/, end: /\)/,
+      contains: ["self", hljs.C_LINE_COMMENT_MODE,hljs.C_BLOCK_COMMENT_MODE]
+    },
```
Comment (egor-rogov) [inline src/languages/javascript.js:96]: What is the purpose of this "parens eater"? Due to this, some highlighting is lost inside params.
E. g. for this snippet:
```
sides.every((length,width=(3+2+(4/5))) => length > 0 );
```
we had before:
```
sides.every(<span class="hljs-function">(<span class="hljs-params">length,width=(
<span class="hljs-number">3</span>+<span class="hljs-number">2</span>+(<span class="hljs-number">4</span>/<span class="hljs-number">5</span></span>))
) =&gt;</span> length &gt; <span class="hljs-number">0</span> );
```
and after the patch we have:
```
sides.every(<span class="hljs-function">(<span class="hljs-params">length,width=(
3+2+(4/5))</span>
) =&gt;</span> length &gt; <span class="hljs-number">0</span> );
```
I. e. numbers are no more highlighted.
(In typescript, numbers are not highlighted even now)

Comment (yyyc514) [reply src/languages/javascript.js:96]: We have to count parens to know which one is the proper one to stop at... but this issue you pointed out can be resolved by just adding more things to the params ruleset. I'll look into that.

Comment (taufik-nurrohman) [reply src/languages/javascript.js:96]: https://github.com/highlightjs/highlight.js/pull/2372#issuecomment-583271131

Comment (yyyc514) [reply src/languages/javascript.js:96]: ?

Comment (taufik-nurrohman) [reply src/languages/javascript.js:96]: @yyyc514 It is okay. It&rsquo;s just that I know this issue will come up as I thought about the function argument pattern problem, like in the comment I have made there.

Comment (yyyc514) [reply src/languages/javascript.js:96]: This fix handles both those cases without issue.

Comment (yyyc514) [reply src/languages/javascript.js]: @egor-rogov Have another look please now.

**Pull request merged**